### PR TITLE
Implementation of toga.Command for Android

### PR DESCRIPTION
**Pull request by t-arn**
This PR implements toga.Command for Android

Only the Commands in the 'Commands' group are supported. 
Commands from other groups should eventually go into the navigation drawer. 

Commands can also be added to the toolbar and will be shown as icons (when available) or as text in the app bar. When there is not enough space in the app bar, the action will 'overflow' into the options menu (see 'TB action 7' with the Tiberius icon in the screenshots below)

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

Screenshot_1636095383

Screenshot_1636095397

Screenshot_1636095415

Screenshot_1636095422

### CODE REVIEW

File: examples/command/command/app.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        # The submenu group has an order that places it in the parent menu.
+        # The items have an explicit ordering that overrides the default
+        # alphabetical ordering.
+        sub_menu = toga.Group("Sub Menu", parent=toga.Group.COMMANDS, order=1)
+        cmd5 = toga.Command(
+            self.action5,
+            label='TB Action 5',
+            tooltip='Perform toolbar action 5',
+            order=2,
+            group=sub_menu
```
Comment (freakboy3742) [inline examples/command/command/app.py:109]: If this command doesn't have an icon deliberately, it would be worth flagging that in the code. My initial impression was "something is broken", and I went looking for why the icon wasn't loading.

File: src/android/toga_android/app.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
 import asyncio
+import toga
 
 from rubicon.java import android_events
 from toga.handlers import wrapped_handler
 
 from .libs.activity import IPythonApp, MainActivity
+from .libs.android.content import Intent
+from .libs.android.net import Uri
```
Comment (freakboy3742) [inline src/android/toga_android/app.py:9]: Are we missing a file here? This PR doesn't define `android.net` AFAICT....

File: src/android/toga_android/app.py [hunk 3]
```diff
@@ -62,6 +67,86 @@ def onActivityResult(self, requestCode, resultCode, resultData):
         except KeyError:
             print("No intent matching request code {requestCode}")
 
+    def onOptionsItemSelected(self, menuitem):
+        consumed = False
+        for cmd in self._impl.interface.commands._commands.union(self._impl.interface.main_window.toolbar._commands):
```
Comment (freakboy3742) [inline src/android/toga_android/app.py:72]: AFAICT, this shouldn't be needed. Any command added to a toolbar should be automatically on the list of all commands; the toolbar is only there as a UI affordance to make the command easily accessible.

Also - is this the callback when a menu item is actually invoked? If it is... why are we iterating over all commands? Shouldn't we be setting up a mapping from `itemId` to `cmd` so we can do an O(1) lookup?

File: src/core/toga/app.py [hunk 4]
```diff
@@ -499,7 +499,10 @@ def visit_homepage(self):
         If the application metadata doesn't define a homepage, this is a no-op.
         """
         if self.home_page is not None:
-            webbrowser.open(self.home_page)
+            if hasattr(self._impl, 'visit_homepage'):
+                self._impl.visit_homepage()
```
Comment (freakboy3742) [inline src/core/toga/app.py:503]: What's the purpose of this change? If the issue is that `webbrowser.open()` doesn't work on Android, then we should fix that in the Python support library. If it's not possible to fix in a generic way, we should make "open_url" a first-class requirement of the App interface, rather than have this sort of "optional" logic.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I've taken a first pass; this definitely looks like it's on the right track. I wasn't able to test it live because it looks like the PR is missing a new module (`toga_android.libs.android.net`); there was also the problem with the VERSION flag in the Gradle template. It might be worth adding that to the beeware/briefcase-android-gradle-template#37 PR; although it's a separate change, it's very much required in order to make this code work with a development version of Toga.

Don't worry too much about the coverage warnings - I'm almost at the point of turning them off for Toga, because they're not really providing any real value. They're useful on a project like Briefcase that has (and wishes to maintain) near complete coverage; but Toga's testing isn't anywhere near as complete - and will be hard to make complete, because GUI features are notoriously difficult to test.

`webbrowser.open()` *should* work on every platform - it definitely works on all the desktop platforms, and I've tested it on iOS (and a patch to the Python source code was required in that case).

**Comment by t-arn** [comment]
@freakboy3742 Hello Russel
Thank you for the fast review! I will address the issues you raised next week.

How is the "resolved" button on the comments meant to work? Do I mark a comment as resolved or does the author of the comment (you) mark it as resolved?

**Comment by freakboy3742** [comment]
Usage of the "Resolved" button is highly variable depending on the project; my preference is that the person who raises the comment is the person who marks it as resolved (since they're the best judge of whether it has been).

File: src/android/toga_android/app.py (hunk 2 above)
Comment (t-arn) [reply src/android/toga_android/app.py:9]: Sorry about this! I *did* test everything and it *did* work. 

I first based the branch on a wrong parent branch and then had to redo all the changes on the new branch commands-for-android. Seems that I missed the `net` module then, but on my file system, the file was still there when I tested it all (again)...

But now, it seems that I don't need it, because no Android specific implementation is needed.

**Comment by t-arn** [comment]
@freakboy3742 Regarding the VERSION problem in the template: yes, I could add a fix to my template PR. I found the place in the code:
```
project​({{ cookiecutter.app_name|upper }} ​VERSION​ {{ cookiecutter.version }})
```
It's the variable `cookiecutter.version` that causes the problem on dev.
But where would I best add the logic for removing the `dev029` suffix?

**Comment by freakboy3742** [comment]
I think the best option is to drop the VERSION tag entirely. I added it in because it I could see it was an available option in CMake, but we're not actually using it anywhere. 

The other alternative would be to do something akin to the `version_triple` tag that is used on Windows MSI installers; Windows has a similar format restriction, so we simplify the version to just the numeric triple. However, this will require a change to both the template *and* briefcase, for no real gain in functionality. If you're feeling especially enthused, then by all means do this... but it's not really worth it IMHO.

File: examples/command/command/app.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+        # Outermost box
+        outer_box = toga.Box(
+            children=[btn_box, self.textpanel],
+            style=Pack(
+                flex=1,
+                direction=COLUMN,
+                padding=10,
+                width=500,
+                height=300
```
[the remark this answers is not in the document]
Comment (t-arn) [reply examples/command/command/app.py:168]: OK, fixed

File: examples/command/command/app.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    def action0(self, widget):
+        print("action 0")
+        self.textpanel.value += 'action 0\n'
+
+    def action1(self, widget):
+        print("action 1")
+        self.textpanel.value += 'action 1\n'
+        for cmd in self.app.commands:
+            self.textpanel.value += repr(cmd)
```
[the remark this answers is not in the document]
Comment (t-arn) [reply examples/command/command/app.py:23]: OK, removed.

File: examples/command/command/app.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    def action6(self, widget):
+        print("action 6")
+        self.textpanel.value += 'action 6\n'
+
+    def action7(self, widget):
+        print("action 7")
+        self.textpanel.value += 'action 7\n'
+
+    def startup(self):
+        # the icons on Android are only found when path is complete
```
[the remark this answers is not in the document]
Comment (t-arn) [reply examples/command/command/app.py:46]: OK, checked again. Relative paths *do* work - the problem I originally had (and interpreted as path resolution problem) is a different one: When an icon is not found (it my case it was command.png which was missing), Toga tries to use the default icon and *that* fails on Android - at least when running from dev:
```
Traceback (most recent call last):
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 54, in bind
    full_path = self._full_path(
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 105, in _full_path
    raise FileNotFoundError(
FileNotFoundError: Can't find icon resources/brutus-256a
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 54, in bind
    full_path = self._full_path(
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 105, in _full_path
    raise FileNotFoundError(
FileNotFoundError: Can't find icon resources/toga
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 54, in bind
    full_path = self._full_path(
  File "/data/user/0/org.beeware.command/files/python/user_code/app_packages/toga/icons.py", line 105, in _full_path
    raise FileNotFoundError(
FileNotFoundError: Can't find icon resources/toga
```
and this endless loop keeps going until `RecursionError: maximum recursion depth exceeded while calling a Python object`

File: examples/command/command/app.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        # brutus_icon_256 = str(self.paths.app) + os.sep + "resources" + os.sep + "brutus-256"
+        brutus_icon_256 = "resources" + os.sep + "brutus-256"
+        cricket_icon_256 = str(self.paths.app) + os.sep + "resources" + os.sep + "cricket-256"
```
[the remark this answers is not in the document]
Comment (t-arn) [reply examples/command/command/app.py:49]: OK, fixed: 
brutus_icon_256 = "resources/brutus-256" **does** work, both on Windows and Android

File: examples/command/command/app.py (hunk 1 above)
Comment (t-arn) [reply examples/command/command/app.py:109]: Yes, this is intentional - I added a commend in the code

File: src/android/toga_android/app.py (hunk 2 above)
Comment (t-arn) [reply src/android/toga_android/app.py:9]: OK, reverted `visit_homepage` back to the original version of toga/app.py:
```
    def visit_homepage(self):
        if self.home_page is not None:
                webbrowser.open(self.home_page)
```
But now, we have the problem, that nothing happens on Android. No webbrowser opens, no error message. In logcat, I only see this:
```
D/MainActivity: onOptionsItemSelected() start
D/Python: Native invocation 3542967056 :: onOptionsItemSelected
D/Python: Native invocation done.
D/MainActivity: onOptionsItemSelected() complete
```

File: src/android/toga_android/app.py (hunk 3 above)
Comment (t-arn) [reply src/android/toga_android/app.py:72]: > AFAICT, this shouldn't be needed.

Right! I'll remove the union

Comment (t-arn) [reply src/android/toga_android/app.py:72]: > Shouldn't we be setting up a mapping from itemId to cmd ?

@freakboy3742 
Yes, that would be the other alternative. I chose to add the itemId as additional attribute of Command.
Would you prefer that I set up a lookup dictionary in `onPrepareOptionsMenu` which we could then use in `onOptionsItemSelected` ?

File: src/core/toga/app.py (hunk 4 above)
Comment (t-arn) [reply src/core/toga/app.py:503]: @freakboy3742 `webbrowser.open` in fact does not work on Android.
I don't know how to fix this in beeware/Python-Android-support - this patching approach is beyond me.
How should we proceed here? Like in `about` where we let the backends implement the appropriate code?
This I could do easily, but I cannot fix `webbrowser.open()` myself.

**Comment by t-arn** [comment]
@freakboy3742 I would like your opinion on following implementation decision I made:
I chose to not add toolbar actions anywhere in the options menu.
The reason is that otherwise, you might end up with the same command being 2 times in the options menu.
This happens when you add the same command to the "Commands" menu and also to the toolbar AND there is not enough space in the toolbar. Then, the toolbar action "spills over" to the options menu and you'll have that action then twice in the options menu which I find really ugly.

I could also have chosen to only prevent the action duplication for the Commands menu and allow a toolbar action to also be in some other menu or sub-menu of "Command". But then, this behavior could be seen as inconsistent by the user which is why I chose the current implementation.

What are your thoughts on this?

File: src/android/toga_android/app.py (hunk 3 above)
Comment (t-arn) [reply src/android/toga_android/app.py]: I'll answer my question myself: It is better to store backend data (menuitem id) in a variable on the implementation layer rather than adding it to an object of the interface layer.
-> using lookup dictionary in the backend now

**Comment by t-arn** [comment]
@freakboy3742 Please re-review this PR, thanks!

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is generally looking pretty good. I've flagged one really minor thing; and I've merged the template change that is needed to support this patch in Briefcase.

There are a couple of open questions:

The handling of `webbrowser.open()` is an awkward one; thinking about it some more, it's going to be difficult to fix at the level of the support package. The fix for iOS involved using ctypes to replicate one specific feature of Rubicon to invoke one native method; unfortunately, the same option doesn't really exist for Android, because there's no built-into-Python way to invoke Java. However, I'm happy to punt on this one - at least partially because of the next point....

This is definitely *a* solution for menus toolbar handling in Android - but I'm not sure it's "right" - at least partially because I don't know what is idiomatic for Android. 

There's essentially 2 questions to answer here:
1. What would the ideal Options menu look like on a well designed Android app - something published by Google, following all their Material style guides
2. How do we map the app description that Toga provides onto that ideal.

So - when it comes to the "built in" menu items:
* Does an ideal app have an About item? If so, where?
* Does an ideal app have a link to the app homepage? If so, where?
* Is there anything else that an ideal app should have on it's option menu?

Essentially, the "hello world" tutorial should turn out an Android app that "does the right thing by default", without the end user needing to do anything. I'm not a native Android user - so I'm not a *great* source of opinion here - but my impression is that a link to the homepage probably isn't needed. 

Anything on top of those "default" menu items comes down to how we choose to interpret menus and toolbars. The key question here: What is the stylistic significance between having an icon on the toolbar, and a menu item in what I think you're referring to as the "overflow" menu? 

In a desktop app, every command appears in the menu, and you can select "important" ones to appear in a toolbar. On a mobile app (especially on a phone), the toolbar real estate is so limited that you've got a pretty hard limit on the number of toolbar items.

If I'm reading this right, you're effectively using the "menu + toolbar" as a unified entity, with "toolbar" really acting as a marker of "important" menu items. I can see how that makes sense; I guess my question is whether there's any *other* interpretation that might make more sense. 

For example, I've considered whether something like a [side menu](https://github.com/jonkykong/SideMenu) would make sense for menus; the icons shown in the toolbar would then become the actual toolbar (with the overflow existing to support apps that put too many items in the toolbar). iOS doesn't provide a side menu out of the box as an API - but it's a really common idiom. Would this approach make sense on Android (albeit one that requires much more than this PR is currently doing - and for the record, I'd be entirely happy with a PR that only tackled the "toolbar" part of this piece, and treated the full menu as later work).

**Comment by t-arn** [comment]
Thank you for your review, @freakboy3742 !

The side menu you mentioned is a very common UI component on Android - it's called the 'navigation drawer' there.
I have also the feeling that the menus should go into the navigation drawer rather than into sub-menus of the options menu. I am not sure though, if we can make the navigation drawer work with beeware...

So, I suggest the following: 
I change this PR to just handle the toolbar as it is and use the options menu for the "Command" menu only and for overflowing toolbar actions. The overflowing of the toolbar actions to the option menu is an automatic behavior of Android and also suggests to view toolbar and options menu as a unified entity.

And for handling the other menus, I will try to make the navigation drawer work in beeware.

Do you agree with this suggestion?

File: src/android/toga_android/app.py [hunk 9]
```diff
@@ -62,6 +66,84 @@ def onActivityResult(self, requestCode, resultCode, resultData):
         except KeyError:
             print("No intent matching request code {requestCode}")
 
+    def onOptionsItemSelected(self, menuitem):
+        consumed = False
+        if menuitem.getItemId() in self.menuitem_mapping:
```
[the remark this answers is not in the document]
Comment (t-arn) [reply src/android/toga_android/app.py:71]: ok, fixed

**Comment by t-arn** [comment]
@freakboy3742 I now reduced the scope of this PR to just handle the toolbar and the commands in the "Commands" group.
Please re-review.

**Review by freakboy3742 — APPROVED** [review]
I've made a couple of minor tweaks (one config change that accidentally leaked into pyproject.toml, and a small cleanup of the logic around the key and path handling for command groups), but otherwise, I think this is good to go.

I'm not 100% convinced we should carve out the "Command" group exception - but that's something we can revisit once we actually implement a full navigation drawer; in the meantime, this approach means commands are "visible by default", which at the very least seems reasonable as an interim approach.

**Comment by t-arn** [comment]
Thank you for the review and the cleanup!
